README
# Standalone javascript image lazy loader

Use it like this :

```html
<img
  data-src="real/image/src.jpg"
  src=data:image/gif;base64,R0lGODlhAQABAID/AP///wAAACwAAAAAAQABAAACAkQBADs=
  onload=lzld(this) />
```

The base 64 src should be the smallest possible, taken from http://www.google-analytics.com/__utm.gif

Do not worry about the size overhead of adding a lot of base 64 src images to your page :
 GZIP is here to help (http://www.gzip.org/deflate.html).

 http://www.fasterize.com


HISTORY.md
# 3.2.2 (2015-12-14)

  * upgrade in-viewport

# 3.2.1 (2015-06-03)

  * upgrade in-viewport, no new feature

# 3.2.0 (2015-05-24)

  * upgrade in-viewport, fixes children in hidden parent case

# 3.1.0 (2015-05-23)

  * feat: upgrade in-viewport to 3.1.0, checks element is visible before triggering callback

# 3.0.2 (2015-03-14)

  * upgrade all deps

# 3.0.1 (2015-03-12)
  
  * switch to closurecompiler

# 3.0.0 (2015-01-30)

  * refactor using browserify



test/404.js
// https://github.com/vvo/lazyload/issues/62
describe('when an image resolves to 404', function() {
  require('./fixtures/bootstrap.js');
  beforeEach(h.clean);
  afterEach(h.clean);

  var fakeSrc = '../b.gif?'+(+new Date());
  var standardSrc = '/IDONOTEXISTSLOL.gif?'+(+new Date());
  var lazyFunc = 'customCallback';
  var calls = 0;

  var test;
  window[lazyFunc] = true;

  beforeEach(function() {
    window[lazyFunc] = lazyload({
      src: function(elt) {
        calls++;
        return elt.getAttribute('data-src');
      }
    });

    test = h.createTest({
      tagName: 'img',
      attributes: {
        id: 'custom-callback',
        src: fakeSrc,
        'data-src': standardSrc,
        width: 10,
        height: 10,
        onload: lazyFunc+'(this)',
        onerror: lazyFunc+'(this)',
      }
    });
    h.insertTest(test);
  });

  describe('after some scrolling', function() {
    beforeEach(h.scroller(0, 50));
    beforeEach(h.wait(200));
    beforeEach(h.scroller(0, 0));
    beforeEach(h.wait(200));

    it('custom callback was called only once (no infinite loop)', function() {
      assert.equal(calls, 1, 'We should have called the callback only once');
    });
  });
});


test/helper.js
(function(win, doc){
  var defaultAttr = 'data-src'

  win.eltNotLoaded = function eltNotLoaded(elt, lazyAttr) {
    lazyAttr = lazyAttr || defaultAttr;

    return function(done) {
      setTimeout(function() {
        assert(elt['data-lzled'] === undefined);
        done();
      }, 25);
    }
  }

  win.eltLoaded = function eltLoaded(elt, lazyAttr) {
    lazyAttr = lazyAttr || defaultAttr;

    return function(done) {
      setTimeout(function() {
        assert(elt['data-lzled'] === true);
        done();
      }, 25);
    }
  }

})(window, document);

test/iframe.js
describe('lazyloading an iframe', function() {
  require('./fixtures/bootstrap.js');
  beforeEach(h.clean);
  afterEach(h.clean);

  var fakeSrc = 'about:blank';
  var realSrc = '/test/fixtures/page.html';

  var test;

  beforeEach(function() {
    test = h.createTest({
      tagName: 'iframe',
      attributes: {
        id: 'iframetest',
        src: fakeSrc,
        'data-src': realSrc,
        width: 10,
        height: 10,
        onload: 'lzld(this)'
      },
      style: {
        top: '5000px'
      }
    });

    h.insertTest(test);
  });

  it('src currently fake', function() {
    assert(test.src.indexOf(fakeSrc) !== -1);
  });

  it('getAttribute still gives real src', function() {
    assert.equal(realSrc, test.getAttribute('src'));
  });

  describe('scrolling to its position', function() {
    mocha.globals(['iframetest', 'navigator']);
    this.timeout(10000);
    beforeEach(h.scroller(0, 2500));
    beforeEach(h.wait(200));
    beforeEach(h.scroller(0, 5000));
    beforeEach(h.wait(600));

    it('loads the iframe', h.eltLoaded('iframetest'));
  });

});

LL.js
(function(window, document){

  function throttle(fn, minDelay) {
    var lastCall = 0;
    return function() {
      var now = +new Date;
      if (now - lastCall < minDelay) {
        return;
      }
      lastCall = now;
      // we do not return anything as
      // https://github.com/documentcloud/underscore/issues/387
      fn.apply(this, arguments);
    }
  }

  // cross browser event add
  function addEvent( el, type, fn ) {
    if (el.attachEvent) {
      el.attachEvent && el.attachEvent( 'on' + type, fn );
    } else {
      el.addEventListener( type, fn, false );
    }
  }

  // cross browser event remove
  function removeEvent( el, type, fn ) {
    if (el.detachEvent) {
      el.detachEvent && el.detachEvent( 'on' + type, fn );
    } else {
      el.removeEventListener( type, fn, false );
    }
  }

  function showIfVisible(img, index) {
    // 200 is the vertical offset used for preloading soon to be visible images
    if (img.getBoundingClientRect().top < winH + 200) {
      img.src = img.getAttribute(lazyAttr);
      img.removeAttribute(lazyAttr);
      // we do not use splice() as we could be in the checkImages's for loop
      imgs[index] = null;

      return true; // img shown
    } else {
      return false; // img to be shown
    }
  }

  function unsubscribe() {
    removeEvent(window, 'resize', getWindowHeightThrottle);
    removeEvent(window, 'scroll', checkImagesThrottle);
  }

  var
    lazyAttr = 'data-src',
    winH,
    imgs = [],
    pageHasLoaded,

    // cross browser window height
    getWindowHeightThrottle = throttle(function getWindowHeight() {
      winH = window.innerHeight ||
        (document.documentElement && document.documentElement.clientHeight) ||
        (document.body && document.body.clientHeight) ||
        10000;
    }, 50),

    checkImagesThrottle = throttle(function checkImages() {
      var
        last = imgs.length,
        current,
        allImagesDone = true;

      for (current = 0; current < last; current++) {
        var img = imgs[current];
        // if showIfVisible is false, it means we have some waiting images to be
        // shown
        if(img !== null && showIfVisible(img, current) === false) {
          allImagesDone = false;
        }
      }

      if (allImagesDone && pageHasLoaded) {
        unsubscribe();
      }
    }, 20);

  // Indicates that the page is optimized
  window['fstrz'] = true;
  // Import any fzns object (Fasterize Name Space)
  window['fzns'] = window['fzns'] || {
    'LL' : {
      's' : function(el) {
        // To avoid onload being called and called and called ...
        el.onload = null;
        showIfVisible(el, imgs.push(el) - 1);
      }
    }
  };

  getWindowHeightThrottle();
  addEvent(window, 'resize', getWindowHeightThrottle);
  addEvent(window, 'scroll', checkImagesThrottle);
  addEvent(window, 'load', function() {
    pageHasLoaded = true;
  });

})(this, document)


Gruntfile.js
module.exports = function(grunt) {
  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),

    concat: {
      build: {
        src:  [
          'node_modules/in-viewport/in-viewport.js',
          '<%= pkg.name %>.js'],
        dest: 'build/<%= pkg.name %>.js'
      }
    },

    'closure-compiler': {
      min: {
        js: 'build/<%= pkg.name %>.js',
        jsOutputFile: 'build/<%= pkg.name %>.min.js',
        options: {
          compilation_level: 'ADVANCED_OPTIMIZATIONS'
        }
      }
    },

    jshint: {
      files: ['<%= pkg.name %>.js', 'package.json'],
      options: {
        jshintrc: '.jshintrc',
        src: ['lazyload.js', 'Gruntfile.js']
      }
    }
  });

  grunt.loadNpmTasks('grunt-closure-compiler');
  grunt.loadNpmTasks('grunt-contrib-concat');
  grunt.loadNpmTasks('grunt-contrib-jshint');

  grunt.registerTask('default', [
    'jshint',
    'concat',
    'closure-compiler',
  ]);

};

components/in-viewport/in-viewport.js
/**
* @license in-viewport v0.3.0 | github.com/vvo/in-viewport#license
*/

(function(win, doc){
  var instances = [];

  win['inViewport'] = inViewport;

  function inViewport(elt, params, cb) {
    var opts = {
      container: doc.body,
      offset: 0
    };

    if (params === undefined || typeof params === 'function') {
      cb = params;
      params = {};
    }

    var container = opts.container = params['container'] || opts.container;
    var offset = opts.offset = params['offset'] || opts.offset;

    for (var i = 0; i < instances.length; i++) {
      if (instances[i].container === container) {
        return instances[i].inViewport(elt, offset, cb);
      }
    }

    return instances[
      instances.push(createInViewport(container)) - 1
    ].inViewport(elt, offset, cb)
  }

  function addEvent( el, type, fn ) {
    if (el.attachEvent) {
      el.attachEvent( 'on' + type, fn );
    } else {
      el.addEventListener( type, fn, false );
    }
  }

  function debounce(func, wait, immediate) {
    var timeout;
    return function() {
      var context = this, args = arguments;
      var later = function() {
        timeout = null;
        if (!immediate) func.apply(context, args);
      };
      var callNow = immediate && !timeout;
      clearTimeout(timeout);
      timeout = setTimeout(later, wait);
      if (callNow) func.apply(context, args);
    }
  }

  // https://github.com/jquery/sizzle/blob/3136f48b90e3edc84cbaaa6f6f7734ef03775a07/sizzle.js#L708
  var contains = document.documentElement.compareDocumentPosition ?
    function( a, b ) {
      return !!(a.compareDocumentPosition( b ) & 16);
    } :
    document.documentElement.contains ?
    function( a, b ) {
      return a !== b && ( a.contains ? a.contains( b ) : false );
    } :
    function( a, b ) {
      while ( (b = b.parentNode) ) {
        if ( b === a ) {
          return true;
        }
      }
      return false;
    };

  function createInViewport(container) {
    var watches = [];
    var scrollContainer = container === doc.body ? win : container;
    var debouncedCheck = debounce(checkElements, 15);

    addEvent(scrollContainer, 'scroll', debouncedCheck);

    if (scrollContainer === win) {
      addEvent(win, 'resize', debouncedCheck);
    }

    function inViewport(elt, offset, cb) {
      if (!contains(doc.documentElement, elt) ||
          !contains(doc.documentElement, container)) {
          return setTimeout(addWatch(elt, offset, cb), 0);
      }

      var eltRect = elt.getBoundingClientRect();
      var containerRect = container.getBoundingClientRect();

      var pos = {
        left: eltRect.left,
        top: eltRect.top
      };

      var viewport = {
        width: offset,
        height: offset
      };

      if (container === doc.body) {
        viewport.width += doc.documentElement.clientWidth;
        viewport.height += doc.documentElement.clientHeight;
      } else {
        pos.left -= containerRect.left;
        pos.top -= containerRect.top;
        viewport.width += container.clientWidth;
        viewport.height += container.clientHeight;
      }

      var visible =
        pos.left >= -offset &&
        pos.left <= viewport.width &&
        pos.top >= -offset &&
        pos.top <= viewport.height;

      if (visible) {
        if (cb) {
          cb(elt);
        } else {
          return true;
        }
      } else {
        if (cb) {
          setTimeout(addWatch(elt, offset, cb), 0);
        } else {
          return false;
        }
      }
    }

    function addWatch(elt, offset, cb) {
      return function() {
        watches.push(function() {
          inViewport(elt, offset, cb, true);
        });
      }
    }

    function checkElements() {
      var cb;
      while(cb = watches.shift()) {
        cb();
      }
    }

    return {
      container: container,
      inViewport: inViewport
    }
  }

})(window, document);

lazyload.js
module.exports = lazyload;

var inViewport = require('in-viewport');
var lazyAttrs = ['data-src'];

global.lzld = lazyload();

// Provide libs using getAttribute early to get the good src
// and not the fake data-src
replaceGetAttribute('Image');
replaceGetAttribute('IFrame');

function registerLazyAttr(attr) {
  if (indexOf.call(lazyAttrs, attr) === -1) {
    lazyAttrs.push(attr);
  }
}

function lazyload(opts) {
  opts = merge({
    'offset': 333,
    'src': 'data-src',
    'container': false
  }, opts || {});

  if (typeof opts.src === 'string') {
    registerLazyAttr(opts.src);
  }

  var elts = [];

  function show(elt) {
    var src = findRealSrc(elt);

    if (src) {
      elt.src = src;
    }

    elt.setAttribute('data-lzled', true);
    elts[indexOf.call(elts, elt)] = null;
  }

  function findRealSrc(elt) {
    if (typeof opts.src === 'function') {
      return opts.src(elt);
    }

    return elt.getAttribute(opts.src);
  }

  function register(elt) {
    // unsubscribe onload
    // needed by IE < 9, otherwise we get another onload when changing the src
    elt.onload = null;
    elt.removeAttribute('onload');

    // https://github.com/vvo/lazyload/issues/62
    elt.onerror = null;
    elt.removeAttribute('onerror');

    if (indexOf.call(elts, elt) === -1) {
      inViewport(elt, opts, show);
    }
  }

  return register;
}

function replaceGetAttribute(elementName) {
  var fullname = 'HTML' + elementName + 'Element';
  if (fullname in global === false) {
    return;
  }

  var original = global[fullname].prototype.getAttribute;
  global[fullname].prototype.getAttribute = function(name) {
    if (name === 'src') {
      var realSrc;
      for (var i = 0, max = lazyAttrs.length; i < max; i++) {
        realSrc = original.call(this, lazyAttrs[i]);
        if (realSrc) {
          break;
        }
      }

      return realSrc || original.call(this, name);
    }

    // our own lazyloader will go through theses lines
    // because we use getAttribute(opts.src)
    return original.call(this, name);
  };
}

function merge(defaults, opts) {
  for (var name in defaults) {
    if (opts[name] === undefined) {
      opts[name] = defaults[name];
    }
  }

  return opts;
}

// http://webreflection.blogspot.fr/2011/06/partial-polyfills.html
function indexOf(value) {
  for (var i = this.length; i-- && this[i] !== value;) {}
  return i;
}
